linked_list/linked_list.py

```diff
@@ -9,71 +9,145 @@ def __init__(self, value, next_node = None):
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        temp = self.head
+        count = 0
+        while temp:
+            count += 1
+            temp = temp.next
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n^2)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if index < 0:
+            return None
+        else:
+            current = self.head
+            count = 0
+            while current:
+                if count == index:
+                    return current.value
+                current = current.next
+                count += 1
+            return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+
+        while current.next != None:
+            current = current.next
+        
+        return current.value
+
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+        if self.head is None:
+            self.head = self.tail = new_node
+        else:
+            self.tail.next = new_node
+            new_node.previous = self.tail
+            self.tail = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        current = self.head
+        if (self.head == None):
+            return None
+        else:
+            # initializing max to the first node value in the linked list
+            max = self.head.value
+            print(max)
+        while (current != None):
+            # if the current node's value is greater than the max
+            # then, replace the value of max_value with the current node's value
+            if max < current.value:
+                max = current.value
+            current = current.next
+        return max
+        
+
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n^2)
+    # Space Complexity: O(1)
     def delete(self, value):

```

Comment:
✨ However you're only looping through the length of the list once, so time complexity would be O(n)